docs/proposals/policy.md

```diff
@@ -0,0 +1,270 @@
+# Policy
+
+## Problem
+Policy extends the existing authorization rules in Kubernetes (https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/authorization.md) to fulfill requirements around extended verb sets, verb specific attributes, role based management of attribute based access control, and more efficient evaluation of authorization status at larger scales.  For discussion about verbs, see: https://github.com/GoogleCloudPlatform/kubernetes/issues/2877.  For discussion about allowing mutation of some fields in a resource and not others, see: https://github.com/GoogleCloudPlatform/kubernetes/issues/2726.
+
+## Use Cases
+ 1.  I want to apply the same set of policy rules to large numbers of projects.
+ 1.  I want to change the policy rules for a role and have it applied to all projects using that role.
+ 1.  I want to have policy changes take effect quickly, without any down time.
+ 1.  I want to have distinct policy rules to control team membership, project metadata, policy rules for roles, and policy rules for users.
+ 1.  OpenShift will ship with a set of default roles: view, edit, and admin
+
+## Exclusions
+ 1. UI (cli or graphical).  This document describes what policy can do, the data required to make policy work, and runtime interactions to fulfill the policy.  It does **not** attempt to describe a UI (cli or graphical) on top of basic resource CRUD.
+ 1. Teams.  This document assumes that a UserInfo object exists and that it has `UserInfo.getUserName() string` and `UserInfo.getTeamNames() []string`.  How that team information is populated and stored is beyond the scope of making effective policy.
+
+  
+## Policy Example
+Out the box, OpenShift provides three roles: view, edit, and admin.  Viewers are able to see all resources.  Editors are able to view and edit all resources, except ones related to policy and membership.  Admins are able to modify anything.  The roles are expressed like this:
+```
+{
+	"kind": "role",
+	"name": "view",
+	"namespace": "global",
+	"rules": [
+		{ "verbs": ["watch", "list", "get"], "resourceKinds": ["*"] },
+	]
+}
+{
+	"kind": "role",
+	"name": "edit",
+	"namespace": "global",
+	"rules": [
+		{ "verbs": ["*"], "resourceKinds": ["*"] }
+		{ "deny": "true", "verbs": ["create", "update", "delete"], "resourceKinds": ["PolicyRuleBinding", "Role", RoleBinding"] }
+	]
+}
+{
+	"kind": "role",
+	"name": "admin",
+	"namespace": "global",
+	"rules": [
+		{ "verbs": ["*"], "resourceKinds": ["*"] }
+	]
+}
+```
+
+The global admin (guy who owns the openshift installation), creates a project `chisel` and assigns `Hubert` as the admin for the project using the predefined role.  The permissions are expressed like this:
+```
+{
+	"kind": "roleBinding",
+	"name": "ProjectAdmins",
+	"namespace": "chisel",
+	"referencesGlobalRole": "true",
```

Comment:
> It feels less magical to specify the exact namespace of the role

While it feels more natural to write that way, it also allows invalid values.  By having a boolean (either local of global), it ensures that no matter what value is present, the RoleBinding itself cannot illegally reference a Role in another namespace.

It is possible to validate the condition, but it feel safer to choose a data structure that makes it completely impossible to enter invalid values.